Improve performance for file reading/writing

mbadams5 opened the pull request:
Pull request overview
---------------------
This pull request grew from its initial scope to something quite large. This PR focused mostly on improving performance when reading and writing files, including Schedule:File, Schedule:File:Shading, EPW weather files, epJSON, and more.

This PR also improves performance in other portions of the code, such as strings to integers and doubles.

Another large change is moving FMT library from 6.1.2 to 8.0.1, which required understanding the custom Fortran formatting code and adapting it to the new FMT api for custom formatting.

The performance improvements can be seen in the tables below:

| HybridZoneModel_shadow113.idf |              |             |                   |
|-------------------------------|--------------|-------------|-------------------|
|                               | Develop (ms) | Branch (ms) | Runtime Reduction |
| Overall                       | 41130        | 4190        | 89.8%             |
| ProcessScheduleInput          | 35350        | 355         | 99.0%             |
| ReadEPlusWeatherForDay        | 964          | 26          | 97.3%             |
| InterpretWeatherDataLine      | 935          | 10          | 98.9%             |
| commonRun                     | 325          | 104         | 68.0%             |
| WriteTimeStampFormatData      | 120          | 80          | 33.3%             |
| WriteNumericData              | 71           | 57          | 19.7%             |
| SetCurrentWeather             | 340          | 176         | 48.2%             |
| ProcessNumber                 | 528          | 1           | 99.8%             |


| HybridZoneModel.idf           |              |             |                   |
|-------------------------------|--------------|-------------|-------------------|
|                               | Develop (ms) | Branch (ms) | Runtime Reduction |
| Overall                       | 4400         | 3490        | 20.7%             |
| ProcessScheduleInput          | 164          | 38          | 76.8%             |
| ReadEPlusWeatherForDay        | 828          | 33          | 96.0%             |
| InterpretWeatherDataLine      | 809          | 10          | 98.8%             |
| commonRun                     | 261          | 97          | 62.8%             |
| WriteTimeStampFormatData      | 104          | 69          | 33.7%             |
| WriteNumericData              | 68           | 62          | 8.8%              |
| SetCurrentWeather             | 273          | 151         | 44.7%             |
| ProcessNumber                 | 36           | 3           | 91.7%             |

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

mbadams5 committed:
Update Milo dtoa and itoa libraries

mbadams5 committed:
Update SQLite library 3.29.0->3.36.0

mbadams5 committed:
Update Valijson ~v0.3 to v0.5

mbadams5 committed:
Update RE2 2019-09-01 -> 2021-08-01

mbadams5 committed:
Update GTest v1.10.0 -> v1.11.0

mbadams5 committed:
Convert to make_unique

mbadams5 committed:
Convert fmt::string_view to std::string_view

mbadams5 committed:
Move FatalError to EnergyPlus.hh

mbadams5 committed:
Upgrade FMT library 6.1.2 -> 8.0.1 (f69a572)

mbadams5 committed:
Update JSON Library 3.7.3 -> 3.10.0 (4336048)

mbadams5 committed:
Faster file reading

mbadams5 committed:
Faster file writing

mbadams5 committed:
Use fmt::ostream instead of std::ostream in certain spots

mbadams5 committed:
Faster string to double/float

mbadams5 committed:
Improve weather file parsing speed

mbadams5 committed:
Assorted improvements

mbadams5 committed:
Improve schedule:file and schedule:file:shading reading

mbadams5 committed:
Improve IdfParser

mbadams5 committed:
Fix unit test build error

mbadams5 committed:
Simplify ConvertInputFormat code and improve performance

mbadams5 committed:
Fix runtime bug in release build

mbadams5 committed:
Fix bugs and update some

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improved_file_reading

mbadams5 committed:
Fix merge conflicts and build errors/warnings

# Conflicts:
#	src/ConvertInputFormat/main.cpp
#	src/EnergyPlus/AirflowNetworkBalanceManager.cc
#	src/EnergyPlus/DaylightingManager.cc
#	src/EnergyPlus/GroundHeatExchangers.cc
#	src/EnergyPlus/HeatBalanceIntRadExchange.cc
#	src/EnergyPlus/InputProcessing/IdfParser.cc
#	src/EnergyPlus/InternalHeatGains.cc
#	src/EnergyPlus/OutputReportTabular.hh
#	src/EnergyPlus/OutputReports.cc
#	tst/EnergyPlus/unit/PVWatts.unit.cc

mbadams5 committed:
Fix linux build errors

amirroth commented on src/EnergyPlus/FileSystem.cc:
```diff
@@ -210,25 +210,42 @@ namespace FileSystem {
+        auto extension = std::string_view(fs::path(filePath).extension().c_str());
+        std::transform(extension.begin(), extension.end(), buffer.begin(), ::toupper);
+        auto const buffer_view = std::string_view(buffer.data());
+        if (buffer_view.compare(".EPJSON") == 0) {
```
Should implement this using `getEnumerationValue` on a `constexpr std::array<std::string_view, FileTypes::Num>` that maps `FileTypes` enum to file extension strings.

mbadams5 replied:
I'm not 100% convinced this is a good pattern for this use case but I implemented it anyways to test it out (I couldn't find any usage of `getEnumerationValue` in the codebase). My main worry is the enum and array getting out of sync but I think I took care of part of the problem with some `static_assert` statements. It does save me a string allocation though.

amirroth replied:
That's fine, we'll change it later. The enumeration and the array getting out of sync is pearl clutching.

mbadams5 replied:
I did say I implemented it right? 😉

amirroth commented on src/EnergyPlus/FileSystem.hh:
```diff
@@ -85,17 +88,45 @@ namespace FileSystem {
-    enum class InputFileType
+    enum class FileTypes
     {
+        Unknown,
```
We generally call this `Unassigned`, and also set it to -1.

mbadams5 replied:
I don't think `Unassigned` makes sense for a file format we don't know, while `Unknown` does make more sense to me. At the end of the day, it doesn't really matter since no user sees this enum.

amirroth replied:
`Unknown` is fine. `-1` is more important, because that is what getEnumerationValue will return if it doesn't find the value.

mbadams5 committed:
Run clang-format

amirroth commented on src/EnergyPlus/FileSystem.hh:
```diff
@@ -85,17 +88,45 @@ namespace FileSystem {
+        MTR,
+        IMF,
+        IDF,
+        last_flat_file_type = IDF
```
We add a last enumeration value `Num` so that `FileTypes::Num` can be used to size arrays and as a loop bound.

mbadams5 replied:
I am unsure how the enumeration values will increment with the features like `last_flat_file_type = IDF`, but it should work the same. That said, I don't envision using this for size arrays or loop bounds unless I implement something like your comment above to use `getEnumerationValue`.

amirroth replied:
`last_flat_file_type = IDF` will continue to work fine. You also don't really need it if you use `BITF_TEST_ANY(BITF(FileType), BITF(TXT) | BITF(MTR) | BITF (IDF) | BITF(IMF))` etc.

amirroth commented on src/EnergyPlus/FileSystem.hh:
```diff
@@ -85,17 +88,45 @@ namespace FileSystem {
+        return t > FileTypes::last_json_type && t <= FileTypes::last_binary_json_type;
+    }
+    inline constexpr bool is_idf_type(FileTypes t) {
+        return t == FileTypes::IDF || t == FileTypes::IMF;
```
Not a giant deal because this will not be called often, but can use the BITF trick.

mbadams5 replied:
Yeah this isn't called often and when these functions are called often, they are part of template functions and thus fully compile time calculations.

amirroth commented on src/EnergyPlus/General.cc:
```diff
@@ -489,50 +489,45 @@ void DetermineDateTokens(EnergyPlusData &state,
+    static constexpr std::array<std::string_view, NumSingleChars>SingleChars{"/", ":", "-"};
+    static constexpr int NumDoubleChars(6);
+    static constexpr std::array<std::string_view, NumDoubleChars> DoubleChars{"ST ", "ND ", "RD ", "TH ", "OF ", "IN "}; // Need trailing spaces: Want thse only at end of words
+    static constexpr std::array<std::string_view, 12> Months{"JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"};
```
Yes please!

Myoldmopar replied:
:+1:

amirroth commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -17269,47 +17258,45 @@ std::string DateToString(int const codedDate) // word containing encoded month,
+        ++Hour;
+        Minute = 0;
+    }
+    switch (Month) {
```
Ay Caramba! `monthStrings[Month-1]` anyone?

mbadams5 replied:
This was a quick conversation from existing code, but I can add a `static constexpr std::array<std::string_view, 12> Months{"JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"};` in and use that.

amirroth commented on src/EnergyPlus/OutputReports.cc:
```diff
@@ -431,31 +431,31 @@ static void WriteDXFCommon(EnergyPlusData &state, InputOutputFile &of, const std
 
     for (int zones = 1; zones <= state.dataGlobal->NumOfZones; ++zones) {
-        print(of, Format_710, format("Zone={}:{}", zones, normalizeName(state.dataHeatBal->Zone(zones).Name)));
+        print<check_syntax(Format_710)>(of, Format_710, fmt::format("Zone={}:{}", zones, normalizeName(state.dataHeatBal->Zone(zones).Name)));
```
What is this fanciness? I'm going to have to learn about this.

mbadams5 replied:
I'm basically being fancy to avoid our custom Fortran Syntax formatter when the format string doesn't include any Fortran syntax ('R', 'S', 'N', 'Z', 'T'). When we have a `static constexpr std::string_view` of a format, we can check that at compile time with:
```cpp
inline constexpr FormatSyntax check_syntax(const std::string_view format_str)
{
    if (is_fortran_syntax(format_str)) {
        return FormatSyntax::Fortran;
    } else {
        return FormatSyntax::FMT;
    }
}
```

The returned enum is used by the various `print` functions in IOFiles.hh to switch at compile time between using normal FMT syntax or our custom Fortran formatter.

It is hard to tell how much time is saved by using the normal FMT syntax vs the custom formatter but there could be a difference for some format strings since the custom formatter sometimes has to call itself.

amirroth replied:
TIL: this!

mbadams5 committed:
Try to fix clang-format again

mbadams5 committed:
Guess I forgot the tst files when formatting

mbadams5 committed:
Implement review changes

mbadams5 committed:
Fix build errors and unit test failures

mbadams5 committed:
Merge remote-tracking branch 'origin/improved_file_reading' into improved_file_reading

mbadams5 committed:
Need to add FromChars.hh

mbadams5 committed:
Merge remote-tracking branch 'origin/improved_file_reading' into improved_file_reading

mbadams5 committed:
Bring in update to allow tuple joins

mbadams5 committed:
Fix bug in fmt library since including ranges.h now

mbadams5 committed:
Fix bugs causing runtime errors

mbadams5 committed:
Partially tested attempt to fix unit test and integration failures

mbadams5 committed:
Fix clang-format error

mbadams5 committed:
Fix various bugs in unit tests

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improved_file_reading

mbadams5 committed:
Fix more bugs

mbadams5 committed:
Fix unit tests
New schedule:file parser is more strict
Changed "z" in LocalEnvData.csv to 0 to match previous implementation
Gtest is more strict about exact matching or something now too for other
unit test

mbadams5 committed:
filesystem::path uses wchar_t on windows but char on everything else
This was causing issues with std::string_view and fmt to some extent
This is a hacky solution for now to just create a string but it would
be good to avoid the string creation if possible

mbadams5 committed:
MSVC didn't like comparing const char * and std::string_view::end()
It wasn't implicitly converting the iterator to a pointer
This uses the data() call which is a const char * to accomplish the same
effect

mbadams5 committed:
Fix build errors and other warnings

mbadams5 committed:
Remove all `using namespace ObjexxFCL;` from unit tests
This was causing build errors on MSVC due to ambiguity in FMT and Objexx

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improved_file_reading

mbadams5 committed:
Fix clang-format issues

mbadams5 committed:
Latest library updates

mbadams5 committed:
Fix bug when no <charconv>

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improved_file_reading

mbadams5 committed:
Fix windows and linux bugs

mbadams5 committed:
Fix bugs and warnings

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improved_file_reading

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improved_file_reading

# Conflicts:
#	src/EnergyPlus/InputProcessing/InputProcessor.cc

mbadams5 committed:
Fix runtime error on AppleClang <12.0.5

mbadams5 committed:
Actually fix runtime error on AppleClang <12.0.5

mbadams5 committed:
Fix windows runtime error

Apparently MSVC doesn't like this being unique_ptr since something, somewhere is terminating the pointer but then tries to delete it again.....? Anyways, changing this to shared_ptr since it won't delete until all references are 0. The original code was a memory leak (although probably not a bad one).

mbadams5 committed:
Fix bug in unit test

mbadams5 committed:
Merge remote-tracking branch 'origin/develop' into improved_file_reading

Myoldmopar commented on scripts/dev/generate_epJSON_schema/idd_parser.py:
```diff
@@ -152,7 +152,7 @@ class Data:
 
 def parse_idd(data):
-    root = {'$schema': "http://json-schema.org/draft-04/schema#", 'properties': {}}
+    root = {'$schema': "https://json-schema.org/draft-07/schema#", 'properties': {}}
```
What are the implications of changing to draft 7?  Does it unlock more capabilities in the schema?

mbadams5 replied:
Yes, it does. There was only 1 minor change needed to update our schema to draft 7 (exclusiveMinimum and exclusiveMaximum are now values instead of boolean, basically you put either minimum or exclusiveMinimum, not both, now).

Myoldmopar commented on scripts/dev/generate_embeddable_epJSON_schema/generate_embeddable_epJSON_schema.cpp:
```diff
@@ -19,7 +19,7 @@ int main( int argc, char const *argv[] )
 	auto const v_cbor = json::to_cbor( input_json );
 
-	printf("	const static std::array< std::uint8_t, %zu > embeddedSchema = {{\n", v_cbor.size() );
+	printf("	static constexpr std::array< std::uint8_t, %zu > embeddedSchema = {{\n", v_cbor.size() );
```
Does this have any impact on performance or is it just more appropriate?

mbadams5 replied:
I don't think it affects performance, however, it can be used in `constexpr` functions now (It likely could have before, but I haven't tested or read to see if that is true)

Myoldmopar commented on src/EnergyPlus/Autosizing/Base.cc:
```diff
@@ -312,9 +312,9 @@ void BaseSizer::reportSizerOutput(EnergyPlusData &state,
 {
 
-    static constexpr fmt::string_view Format_990(
+    static constexpr std::string_view Format_990(
```
:+1:

Myoldmopar commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -798,7 +798,7 @@ namespace AirLoopHVACDOAS {
                 AirLoopMixer thisAirLoopMixer;
                 thisDOAS.m_CompPointerAirLoopMixer = thisAirLoopMixer.factory(state, thisDOAS.m_AirLoopMixerIndex, thisDOAS.AirLoopMixerName);
-                thisDOAS.AirLoopSplitterName = UtilityRoutines::MakeUPPERCase(AsString(fields.at("airloophvac_splitter_name"))); //
+                thisDOAS.AirLoopSplitterName = UtilityRoutines::MakeUPPERCase(fields.at("airloophvac_splitter_name").get<std::string>()); //
```
Much nicer.

Myoldmopar commented on src/EnergyPlus/CurveManager.cc:
```diff
@@ -2385,21 +2387,18 @@ namespace CurveManager {
+                    str.remove_prefix(first_char);
+                }
+                double result = 0;
+                auto answer = fast_float::from_chars(str.data(), str.data() + str.size(), result);
```
:+1:

Myoldmopar commented on src/EnergyPlus/DataOutputs.hh:
```diff
@@ -78,8 +78,8 @@ namespace DataOutputs {
-        RE2 *pattern;
-        RE2 *case_insensitive_pattern;
+        std::shared_ptr<RE2> pattern = nullptr;
+        std::shared_ptr<RE2> case_insensitive_pattern = nullptr;
```
This will be helpful :+1:

Myoldmopar commented on src/EnergyPlus/FileSystem.hh:
```diff
@@ -148,6 +191,122 @@ namespace FileSystem {
+
+    template <FileTypes fileType> std::string getJSON(const nlohmann::json &data, int const indent = 4)
+    {
+        if constexpr (is_json_type(fileType)) {
```
Nice.

Myoldmopar reviewed (approved):
This is littered with great stuff.  There are 836 files changed, so obviously I'm not going to be able to go line by line.  The changes I have looked over look great though, and CI is very happy.  I think this is ready to go.
